File: 백준/Platinum/7626. 직사각형/직사각형.py
import sys, math
read = sys.stdin.readline
INF = float('inf')

N = int(read())
x_dict, x_list, reg = dict(), set(), list()

for _ in range(N):
    x1, x2, y1, y2 = map(int, read().split())
    reg.append((y1, x1, x2, 1)), reg.append((y2, x1, x2, -1))
    x_list.add(x1), x_list.add(x2)
    
reg.sort()
x_list = sorted(x_list)
nx = len(x_list)
n = 2 ** math.ceil(math.log2(nx - 1))
tree, count = [0] * (2 * n), [0] * (2 * n)

for i in range(nx):
    x_dict[x_list[i]] = i

def update(flag, x1, x2, pos = 1, left = 0, right = n - 1):
    
    mid = (left + right) // 2
    
    if x1 > right or x2 <= left:
        return
    
    elif x1 <= left and x2 > right:
        count[pos] += flag
        
    else:
        if x1 <= mid:
            update(flag, x1, x2, pos*2, left, mid)
        if mid + 1 < x2:
            update(flag, x1, x2, pos*2+1, mid+1, right)
            
    if count[pos] != 0:
        tree[pos] = x_list[right + 1] - x_list[left]
        
    elif left == right:
        tree[pos] = 0
        
    else:
        tree[pos] = tree[pos*2] + tree[pos*2+1]
        
    return

ans = 0
p_y = reg[0][0]

for y, x1, x2, flag in reg:
    if y > p_y:
        ans += tree[1] * (y - p_y)
    update(flag, x_dict[x1], x_dict[x2])
    p_y = y
    
print(ans)

File: 백준/Platinum/7626. 직사각형/직사각형.cc
#include<stdio.h>
#include<algorithm>
using namespace std;

int N;

class Point{
        public:
        int val;
        int idx;
        int y1;
        int y2;

        bool operator<(Point a) const{
                return this->val < a.val;
        }
};
class Y{
        public:
        int val;
        int idx;

        bool operator<(Y a) const{
                return this->val < a.val;
        }
};

Point x[400002];
Y y[400002];

long long int cnt[1600000];
long long int sum[1600000];
class Tree{
        public:
        Tree(){
                for(int i=0;i<N*8;i++){
                        cnt[i] = 0;
                        sum[i] = 0;
                }

        }
        void update(int left, int right, int start, int end, int k, int val){
                int mid = (start+end)/2;
                int ret = 0;
                if(start >=left && end <= right){
                                cnt[k] += val;
                }else{
                        if(left<=mid) update(left, right, start, mid, k*2, val);
                        if(right>mid) update(left, right, mid+1, end, k*2+1, val);
                }

                if(cnt[k]){
                        sum[k] = y[end+1].val - y[start].val;
                }else{
                        if(start==end) sum[k] = 0;
                        else sum[k] = sum[k*2] + sum[k*2+1];
                }
        }
};
int main(){
        long long int ANS = 0;
        scanf("%d",&N);
        for(int i=0; i<N;i++){
                int a = 2*i, b = 2*i+1;
                scanf("%d %d %d %d",&x[a].val, &x[b].val, &y[a].val, &y[b].val);
                x[a].idx = y[a].idx = a;
                x[b].idx = y[b].idx = b;
        }
        sort(y,y+2*N);

                //Coordinate compression
        for(int i=0;i<2*N;i++){
                int arr_idx = y[i].idx / 2;
                if(y[i].idx %2==0){
                        x[arr_idx*2].y1 = i;
                        x[arr_idx*2+1].y1 = i;
                }else{
                        x[arr_idx*2].y2 = i;
                        x[arr_idx*2+1].y2 = i;
                }
        }

        sort(x,x+2*N);
        for(int i=0;i<2*N;i++){
        }

        Tree *t = new Tree();
        for(int i=0;i<2*N;i++){
                if(i>0){
                        ANS += sum[1] * (x[i].val - x[i-1].val);
                }
                if(x[i].idx % 2 ==0){ //Rectangle start
                        t->update(x[i].y1, x[i].y2-1, 0, 2*N-1,1, 1);
                }else{//Rectangle end
                        t->update(x[i].y1, x[i].y2-1, 0, 2*N-1,1, -1);
                }

        }
        printf("%lld\n",ANS);
        return 0;
}